src/main/java/seedu/address/logic/commands/UndoCommand.java

```diff
@@ -26,82 +23,92 @@ public class UndoCommand extends Command {
-            Person personToRemove = addCommand.getToAdd();
-            model.deletePerson(personToRemove);
-            pastCommands.remove(pastCommands.size() - 1);
-            resultMessage = String.format(MESSAGE_UNDO_ADD, personToRemove.getName());
+            isAction = true;
+            resultMessage = addCommand.undo(model, pastCommands);
             break;
         case "edit":
             EditCommand editCommand = (EditCommand) latestCommand;
-            Person bfrEdit = editCommand.getUneditedPerson();
-            Person afterEdit = editCommand.getEditedPerson();
-            model.setPerson(afterEdit, bfrEdit);
-            pastCommands.remove(pastCommands.size() - 1);
-            resultMessage = String.format(MESSAGE_UNDO_EDIT, bfrEdit.getName());
+            isAction = true;
+            resultMessage = editCommand.undo(model, pastCommands);
             break;
 
         case "delete":
             DeleteCommand dltCommand = (DeleteCommand) latestCommand;
-            List<Person> personsToAddBack = dltCommand.getPersonsToDelete();
-
-            for (int i = 0; i < personsToAddBack.size(); i++) {
-                model.addPerson(personsToAddBack.get(i), dltCommand.getTargetIndexes()[i].getZeroBased());
-            }
-            String namesAddedBack = personsToAddBack.stream()
-                    .map(person -> person.getName().toString())
-                    .collect(Collectors.joining(", "));
-            pastCommands.remove(pastCommands.size() - 1);
-            resultMessage = String.format(MESSAGE_UNDO_DELETE, namesAddedBack);
+            isAction = true;
+            resultMessage = dltCommand.undo(model, pastCommands);
             break;
 
         case "clear":
             ClearCommand clearCommand = (ClearCommand) latestCommand;
-            model.setAddressBook(clearCommand.getModel().getAddressBook());
-            model.setUserPrefs(clearCommand.getModel().getUserPrefs());
-            pastCommands.remove(pastCommands.size() - 1);
-            resultMessage = MESSAGE_UNDO_CLEAR;
+            isAction = true;
+            resultMessage = clearCommand.undo(model, pastCommands);
             break;
+
+        case "adda":
+            AddAppointmentCommand addAppointmentCommand = (AddAppointmentCommand) latestCommand;
+            isAction = true;
+            resultMessage = addAppointmentCommand.undo(model, pastCommands);
+            break;
+
+        case "deletea":
+            DeleteAppointmentCommand deleteAppointmentCommand = (DeleteAppointmentCommand) latestCommand;
+            isAction = true;
+            resultMessage = deleteAppointmentCommand.undo(model, pastCommands);
+            break;
+
         default:
-            pastCommands.remove(pastCommands.size() - 1);
+            pastCommands.remove();
             break;
 
         }
-        return new CommandResult(String.format(MESSAGE_UNDO_COMMAND_SUCCESS, resultMessage));
+        if (isAction) {
```

Comment:
Instead of having the `isAction flag`, perhaps can consider returning command result directly in each case? Can also look into the possibility of combining cases using commas, e.g. `case "add", "delete", edit":` to avoid repeating code.